### DISCUSSION TITLE
Can we make more components available in import libraries?

### DISCUSSION THREAD

**Discussion by frederick-vs-ja** [microsoft/STL#3835] [Ideas] [answered]
I found that there were mirrorings of `basic_string`, `basic_string_view`, and `expected` in [`print.cpp`](https://github.com/microsoft/STL/blob/40640c6df9754ea031f8df7c720b89714d19df97/stl/src/print.cpp) during implementing WG21-P2093R14 because the corresponding headers were non-core.

Would it be better to make some headers/classes "partially core" so that we can directly use them in import libraries?

**Comment by cpplearner** [comment, marked as answer]
IIUC that will be ABI-incompatible with code built with e.g. `/MD /D_ITERATOR_DEBUG_LEVEL=1`.

**Reply by StephanTLavavej** [reply]
Yep. Any machinery injected into the STL's import lib (for which we have only 2: release and debug) is effectively **statically linked** to the user's code. Therefore, such machinery cannot depend on `_ITERATOR_DEBUG_LEVEL`, which is why keeping `basic_string` out of the import lib is so critical. Core vs. non-core has been a reasonable way to enforce this restriction, and generally ensure that we keep complicated machinery out of the import lib. (For example, while `basic_string_view` doesn't have a varying representation, it has a lot going on, including `_CONTAINER_DEBUG_LEVEL`, and we'd prefer to avoid mix-and-match headaches with it.)
